src/shogun/machine/GLM.cpp

```diff
@@ -0,0 +1,262 @@
+{
+	SGVector<float64_t> z = compute_z(X, w, l_bias);
+	SGVector<float64_t> result = non_linearity(z);
+	return result;
+}
+
+const SGVector<float64_t> GLM::compute_z(const SGMatrix<float64_t> X, const SGVector<float64_t> w, const float64_t l_bias)
+{
+	SGVector<float64_t> z = linalg::matrix_prod(X, w, false);
+	return z;
+}
+
+const SGVector<float64_t> GLM::non_linearity(const SGVector<float64_t> z)
+{
+	SGVector<float64_t> result;
+	switch (distribution)
+	{
+	case POISSON:
+		result = SGVector<float64_t>(z);
+		float64_t l_bias = 0;
+
+		if(LinearMachine::m_compute_bias)
+			l_bias = (1 - m_eta) * std::exp(m_eta);
+
+		for (int i = 0; i < z.vlen; i++)
+		{
+			if(z[i]>m_eta)
+				result[i] = z[i] * std::exp(m_eta) + l_bias;
+			else
+				result[i] = std::exp(z[i]);
+		}
+		break;
+	
+	default:
+		error("Not a valid distribution type");
+		break;
+	}
+	return result;
+}
+
+const SGVector<float64_t> GLM::gradient_non_linearity(const SGVector<float64_t> z)
+{
+	SGVector<float64_t> result;
+	switch (distribution)
+	{
+	case POISSON:
+		result = SGVector<float64_t>(z);
+		for (int i = 0; i < z.vlen; i++)
+		{
+			if(z[i]>m_eta)
+				
+				result[i] = std::exp(m_eta);
+			else
+				result[i] = std::exp(z[i]);
+		}
+		break;
+	
+	default:
+		error("Not a valid distribution type");
+		break;
+	}
+	return result;
+}
+
+const SGVector<float64_t> GLM::compute_grad_L2_loss_w(const SGMatrix<float64_t> X, const SGVector<float64_t> y, const SGVector<float64_t> w, const float64_t l_bias)
+{
+	auto n_samples = y.vlen;
+	auto n_features = X.num_rows;
+	
+	SGVector<float64_t> z = compute_z(X, w, l_bias);
+	SGVector<float64_t> mu = non_linearity(z);
+	SGVector<float64_t> grad_mu = gradient_non_linearity(z);
+
+	SGVector<float64_t> grad_w(w.vlen);
+	switch (distribution)
+	{
+	case POISSON:
+		SGVector<float64_t> a = y * grad_mu / mu;
+		linalg::transpose_matrix(linalg::add(linalg::matrix_prod(SGMatrix(grad_mu), X, true, false)), (linalg::matrix_prod(SGMatrix(a), X, true, false)), 1, -1);
+		break;
+	
+	default:
+		error("Not a valid distribution type");
+		break;
+	}
+	for (int i = 0; i < grad_w.vlen; i++)
+		grad_w[i] /= n_samples;
+	
+	return grad_w;
+}
+
+const float64_t GLM::compute_grad_L2_loss_bias(const SGMatrix<float64_t> X, const SGVector<float64_t> y, const SGVector<float64_t> w, const float64_t l_bias)
+{
+	auto n_samples = y.vlen;
+	SGVector<float64_t> z = compute_z(X, w, l_bias);
+	SGVector<float64_t> mu = non_linearity(z);
+	SGVector<float64_t> grad_mu = gradient_non_linearity(z);
+
+	float64_t grad_bias = 0;
+	switch (distribution)
+	{
+	case POISSON:
+		for (int i = 0; i < grad_mu.vlen; i++)
+		{
+			grad_bias += grad_mu[i];
+			grad_bias -= y[i] * grad_mu[i] / mu[i];
+		}
+		break;
+	
+	default:
+		error("Not a valid distribution type");
+		break;
+	}
+	grad_bias /= n_samples;
+
+	return grad_bias;
+}
+
+bool GLM::train_machine(std::shared_ptr<Features> data = NULL)
+{
+	std::shared_ptr<SGDMinimizer> minimizer;
```

Comment:
@karlnapf this is what you meant right?
I guess all I have to do now is make a new class that derives from `FirstOrderCostFunction` and then make an object of that class which I will have to pass on to the `minimizer` object so that it can minimize the custom cost function.

Where should I make this class(as in where should it fit in the file structure)?

And should there be different classes for different distributions or just one which gives has a switch-case like structure inside which decides the course of action depending on the `distribution`(`GLMDistribution` enum).

Reply:
is this still an open question?
I am thinking that maybe this full blown approach for the optimization might be overkill. Although it is nice to offer users to parametrise this. Did you get it to work?

Reply:
No, It isn't working as yet. The issue I am facing is that in the `SGDMinimizer` class, it only minimizes one variable. So how should I deal with the weights and gradient separately?
Or should I call the `minimize()` twice, although I am not sure how optimal that would be?

Or otherwise, I can go back to the old `IterativeMachine` approach

Reply:
What do you mean it minimizes only one variable?

Reply:
btw there is also https://github.com/shogun-toolbox/shogun-gpl/blob/master/shogun/optimization/NLOPTMinimizer.h

Reply:
> What do you mean it minimizes only one variable?

I meant it minimizes the cost by changing only one variable, (ie the vector of weights, `m_w`). My bad for the confusion

This is the code for the minimize() function as defined in the `SGDMinimizer` class:
```
float64_t SGDMinimizer::minimize()
{
	init_minimization();

	SGVector<float64_t> variable_reference=m_fun->obtain_variable_reference();
	auto fun= m_fun->as<FirstOrderStochasticCostFunction>();
	require(fun,"the cost function must be a stochastic cost function");
	for(;m_cur_passes<m_num_passes;m_cur_passes++)
	{
		fun->begin_sample();
		while(fun->next_sample())
		{
			m_iter_counter++;
			float64_t learning_rate=1.0;
			if(m_learning_rate)
				learning_rate=m_learning_rate->get_learning_rate(m_iter_counter);
			SGVector<float64_t> grad=m_fun->get_gradient();
			update_gradient(grad,variable_reference);
			m_gradient_updater->update_variable(variable_reference,grad,learning_rate);

			do_proximal_operation(variable_reference);
		}
	}
	float64_t cost=m_fun->get_cost();
	return cost+get_penalty(variable_reference);
}
```
what this does is minimize cost by changing the SGVector `variable_reference`(here `m_w` the vector of weights) using gradient descend. (by calling the `update_gradient()` method)

What do I do if I want it to change the `bias` as well?
The cost function for the bias and the weights are different as we don't apply the elastic net penalty to the bias

Reply:
ah I see now.
That is a problem indeed. You would need to write some code that combines the cost/variables into one to use the framework (or refactor it).
Is that tricky? If so, then we might go back to the old way of doing this?

Reply:
I think that refactoring will make the code hard to read as the variables will become a little ambiguous.

What I think I'll do is use the `IterativeMachine` class to run the iterations but I will still be using the following optimization classes:
`GradientDescendUpdater`
`ConstLearningRate`
`ElasticNetPenalty`

I think in this way the code will remain readable and cleaner. Is this okay?

Reply:
And then have two instances of those updaters?
I am not sure I agree about it being hard to read, if you keep the code as is and only add a bridge to pass it to the optimizer. But OK, go ahead with what you think is best and then we can look at the code and discuss.

Reply:
Can I resolve this?